junifer/preprocess/base.py

```diff
@@ -0,0 +1,174 @@
+
+
+from ..pipeline import PipelineStepMixin
+from ..utils import logger, raise_error
+
+
+class BasePreprocessor(ABC, PipelineStepMixin):
+    """Provide abstract base class for all preprocessors.
+
+    Parameters
+    ----------
+    on : str or list of str, optional
+        The kind of data to apply the preprocessor to. If None,
+        will work on all available data (default None).
+    """
+
+    def __init__(
+        self,
+        on: Optional[Union[List[str], str]] = None,
+    ) -> None:
+        if on is None:
+            on = self.get_valid_inputs()
+        if not isinstance(on, list):
+            on = [on]
+
+        if any(x not in self.get_valid_inputs() for x in on):
+            name = self.__class__.__name__
+            wrong_on = [x for x in on if x not in self.get_valid_inputs()]
+            raise ValueError(f"{name} cannot be computed on {wrong_on}")
+        self._on = on
+
+    def validate_input(self, input: List[str]) -> None:
+        """Validate input.
+
+        Parameters
+        ----------
+        input : list of str
+            The input to the pipeline step. The list must contain the
+            available Junifer Data dictionary keys.
+
+        Raises
+        ------
+        ValueError
+            If the input does not have the required data.
+        """
+        if not any(x in input for x in self._on):
+            raise_error(
+                "Input does not have the required data."
+                f"\t Input: {input}"
+                f"\t Required (any of): {self._on}"
+            )
+
+    @abstractmethod
+    def get_output_kind(self, input: List[str]) -> List[str]:
+        """Get output kind.
+
+        Parameters
+        ----------
+        input : list of str
+            The input to the preprocessor. The list must contain the
+            available Junifer Data dictionary keys.
+
+        Returns
+        -------
+        list of str
+            The updated list of output kinds, as storage possibilities.
+
+        """
+        raise_error(
+            msg="Concrete classes need to implement get_output_kind().",
+            klass=NotImplementedError,
+        )
+
+    def get_valid_inputs(self) -> List[str]:
+        """Get valid data types for input.
+
+        Returns
+        -------
+        list of str
+            The list of data types that can be used as input for this
+            preprocessor.
+        """
+        raise_error(
+            msg="Concrete classes need to implement get_valid_inputs().",
+            klass=NotImplementedError,
+        )
+
+    def get_meta(self, kind: str) -> Dict:
+        """Get metadata.
+
+        Parameters
+        ----------
+        kind : str
+            The kind of pipeline step.
+
+        Returns
+        -------
+        dict
+            The metadata as a dictionary with the only key 'preprocess'.
+        """
+        s_meta = super().get_meta()
+        return {"preprocess": s_meta}
```

Comment:
Should we have a `name` key in `s_meta` with the preprocess's name (`self.name`)?

Reply:
As explained before, the name is "preprocess".